backend/src/main/scala/cromwell/backend/WorkflowBackendActor.scala

```diff
@@ -0,0 +1,83 @@
+package cromwell.backend
+
+import akka.actor.{ActorRef, ActorLogging, Actor}
+import akka.event.LoggingReceive
+import cromwell.backend.WorkflowBackendActor._
+import cromwell.core.CallOutputs
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.{Failure, Success}
+
+object WorkflowBackendActor {
+
+  // Commands
+  sealed trait WorkflowBackendActorMessage
+
+  case class Execute(jobDescriptor: BackendJobDescriptor) extends WorkflowBackendActorMessage
+  case class Recover(jobDescriptor: BackendJobDescriptor) extends WorkflowBackendActorMessage
+  case class Abort(jobDescriptor: BackendJobDescriptor) extends WorkflowBackendActorMessage
+
+  // Events
+  sealed trait WorkflowBackendActorResponse
+
+  sealed trait ExecutionResponse extends WorkflowBackendActorResponse
+  case class ExecutionSucceeded(jobDescriptor: BackendJobDescriptor, callOutputs: CallOutputs) extends ExecutionResponse
+  case class ExecutionFailed(jobDescriptor: BackendJobDescriptor, throwable: Throwable) extends ExecutionResponse
+  case class ExecutionFailedRetryable(jobDescriptor: BackendJobDescriptor, throwable: Throwable) extends ExecutionResponse
+
+  sealed trait AbortResponse extends WorkflowBackendActorResponse
+  case class AbortSucceeded(jobDescriptor: BackendJobDescriptor) extends AbortResponse
+  case class AbortFailed(jobDescriptor: BackendJobDescriptor, throwable: Throwable) extends AbortResponse
+}
+
+/**
+  * Workflow-level actor for executing, recovering and aborting jobs.
+  */
+trait WorkflowBackendActor extends Actor with ActorLogging {
+
+  implicit def ec: ExecutionContext
+  def workflowDescriptor: BackendWorkflowDescriptor
+
+  def receive: Receive = LoggingReceive {
+    case Execute(jobDescriptor) => executeAndReply(execute, jobDescriptor)
+    case Recover(jobDescriptor) => executeAndReply(recover, jobDescriptor)
+    case Abort(jobDescriptor) => abortAndReply(abort, jobDescriptor)
+  }
+
+  private def executeAndReply(execution: BackendJobDescriptor => Future[ExecutionResponse],
+                              jobDescriptor: BackendJobDescriptor,
+                              sndr: ActorRef = sender) = {
+    execution(jobDescriptor) onComplete {
+      case Success(executionResponse: ExecutionResponse) =>
+        sndr ! executionResponse
+      case Failure(exception: Throwable) =>
+        sndr ! ExecutionFailed(jobDescriptor, exception)
+    }
+  }
+
+  private def abortAndReply(abortFunction: BackendJobDescriptor => Future[AbortResponse],
+                            jobDescriptor: BackendJobDescriptor,
+                            sndr: ActorRef = sender) = {
+    abortFunction(jobDescriptor) onComplete {
+      case Success(abortResponse: AbortResponse) =>
+        sndr ! abortResponse
+      case Failure(exception: Throwable) =>
+        sndr ! AbortFailed(jobDescriptor, exception)
+    }
+  }
```

Comment:
I knew someone would feel compelled to improve that eventually... I didn't expect it so quickly!